practical-penguins/trivia_tavern/static/jquery.formset.js

```diff
@@ -0,0 +1,195 @@
+;(function($) {
+    $.fn.formset = function(opts)
+    {
+        var options = $.extend({}, $.fn.formset.defaults, opts),
+            flatExtraClasses = options.extraClasses.join(' '),
+            totalForms = $('#id_' + options.prefix + '-TOTAL_FORMS'),
+            maxForms = $('#id_' + options.prefix + '-MAX_NUM_FORMS'),
+            childElementSelector = 'input,select,textarea,label,div',
+            $$ = $(this),
+
+            applyExtraClasses = function(row, ndx) {
+                if (options.extraClasses) {
+                    row.removeClass(flatExtraClasses);
+                    row.addClass(options.extraClasses[ndx % options.extraClasses.length]);
+                }
+            },
+
+            updateElementIndex = function(elem, prefix, ndx) {
+                var idRegex = new RegExp(prefix + '-(\\d+|__prefix__)-'),
+                    replacement = prefix + '-' + ndx + '-';
+                if (elem.attr("for")) elem.attr("for", elem.attr("for").replace(idRegex, replacement));
+                if (elem.attr('id')) elem.attr('id', elem.attr('id').replace(idRegex, replacement));
+                if (elem.attr('name')) elem.attr('name', elem.attr('name').replace(idRegex, replacement));
+            },
+
+            hasChildElements = function(row) {
+                return row.find(childElementSelector).length > 0;
+            },
+
+            showAddButton = function() {
+                return maxForms.length == 0 ||   // For Django versions pre 1.2
+                    (maxForms.val() == '' || (maxForms.val() - totalForms.val() > 0))
+            },
+
+            insertDeleteLink = function(row) {
+                if (row.is('TR')) {
+                    // If the forms are laid out in table rows, insert
+                    // the remove button into the last table cell:
+                    row.children(':last').append('<a class="' + options.deleteCssClass +'" href="javascript:void(0)">' + options.deleteText + '</a>');
```

Comment:
Although the JavaScript isn't a core part of the review, I did want to say that this part was well commented and written in a reasonably easy to understand manner. There were some areas such as `updateExtraElements()` that tried to do a bit much in single lines (although, I'll admit that style is somewhat common in the JS world), and there could have been a better split for separating function calls from inserted templates. E.g. instead of:

```js
// You can probably find a better name than `temp`, it's just a rough example
var temp = '<a class="' + options.deleteCssClass +'" href="javascript:void(0)">' + options.deleteText + '</a>';
row.children(':last').append(temp);
```

I find that the above format makes the core business logic of the code much easier to extract at a glance, which can be very important for the next person that has to work on it (or yourself, a few months down the road). You can apply this to several areas of the JS code below.

Reply:
> 
> 
> Although the JavaScript isn't a core part of the review, I did want to say that this part was well commented and written in a reasonably easy to understand manner. There were some areas such as `updateExtraElements()` that tried to do a bit much in single lines (although, I'll admit that style is somewhat common in the JS world), and there could have been a better split for separating function calls from inserted templates. E.g. instead of the above:
> 
> ```js
> // You can probably find a better name than `temp`, it's just a rough example
> var temp = '<a class="' + options.deleteCssClass +'" href="javascript:void(0)">' + options.deleteText + '</a>';
> row.children(':last').append(temp);
> ```
> 
> I find that the above format makes the core business logic of the code much easier to extract at a glance, which can be very important for the next person that has to work on it (or yourself, a few months down the road). You can apply this to several areas of the JS code below.